hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go

```diff
@@ -0,0 +1,565 @@
+	if depth > 0 {
+		results = append(results, assignmentHandler(destination, actualDestination))
+	}
+
+	return results
+}
+
+// convertComplexMapProperty handles conversion for map properties with complex values.
+// This function panics if the map keys are not primitive types.
+// This function generates code that looks like this:
+// 	if <source> != nil {
+//		<destination> = make(map[<destinationType.KeyType()]<destinationType.ValueType()>)
+//		for key, value := range <source> {
+// 			<code for producing result from destinationType.ValueType()>
+//			<destination>[key] = <result>
+//		}
+//	}
+func (builder *convertFromArmBuilder) convertComplexMapProperty(
+	source ast.Expr,
+	destination ast.Expr,
+	destinationType *astmodel.MapType,
+	_ string,
+	conversionContext []astmodel.Type,
+	assignmentHandler func(lhs ast.Expr, rhs ast.Expr) ast.Stmt) []ast.Stmt {
+
+	if _, ok := destinationType.KeyType().(*astmodel.PrimitiveType); !ok {
+		panic(fmt.Sprintf("map had non-primitive key type: %v", destinationType.KeyType()))
+	}
+
+	depth := countArraysAndMapsInConversionContext(conversionContext)
+
+	keyIdent := ast.NewIdent("key")
+	valueIdent := ast.NewIdent("value")
+	elemIdent := ast.NewIdent("elem")
+
+	actualDestination := destination // TODO: improve name
+	makeMapToken := token.ASSIGN
+	if depth > 0 {
+		actualDestination = elemIdent
+		elemIdent = ast.NewIdent(fmt.Sprintf("elem%d", depth))
+		makeMapToken = token.DEFINE
+	}
+
+	handler := func(lhs ast.Expr, rhs ast.Expr) ast.Stmt {
+		return astbuilder.InsertMap(lhs, keyIdent, rhs)
+	}
+
+	keyTypeAst := destinationType.KeyType().AsType(builder.codeGenerationContext)
+	valueTypeAst := destinationType.ValueType().AsType(builder.codeGenerationContext)
+
+	makeMapStatement := astbuilder.SimpleAssignment(
+		actualDestination,
+		makeMapToken,
+		astbuilder.MakeMap(keyTypeAst, valueTypeAst))
+	rangeStatement := &ast.RangeStmt{
+		Key:   keyIdent,
+		Value: valueIdent,
+		X:     source,
+		Tok:   token.DEFINE,
+		Body: &ast.BlockStmt{
+			List: builder.fromArmComplexPropertyConversion(
+				valueIdent,
+				actualDestination,
+				destinationType.ValueType(),
+				elemIdent.Name,
+				append(conversionContext, destinationType),
+				handler),
+		},
+	}
+
+	result := &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  source,
+			Op: token.NEQ,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				makeMapStatement,
+				rangeStatement,
+			},
+		},
+	}
+
+	if depth > 0 {
+		result.Body.List = append(result.Body.List, assignmentHandler(destination, actualDestination))
+	}
+
+	return []ast.Stmt{result}
+}
+
+// convertComplexTypeNameProperty handles conversion of complex TypeName properties.
+// This function generates code that looks like this:
+//	<nameHint> := <destinationType>{}
+//	err = <nameHint>.FromArm(owner, <source>)
+//	if err != nil {
+//		return err
+//	}
+//	<destination> = <nameHint>
+func (builder *convertFromArmBuilder) convertComplexTypeNameProperty(
```

Comment:
Yep. Templates sorta make some things easier but make other things harder. Especially in this case I think that there's just some inherent complexity in what we're doing and it's going to be difficult to make it really simple, because the whole notion of "convert an arbitrary object, map, or slice to another object/map or slice" is just a bit fraught.